### PR TITLE
CORDA-1115 - Cannot serialize private nested objects

### PR DESCRIPTION
**Pull request by fenryka**

### CODE REVIEW

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt [hunk 1]
```diff
@@ -513,3 +513,34 @@ fun ClassWhitelist.hasAnnotationInHierarchy(type: Class<*>): Boolean {
             || type.interfaces.any { hasAnnotationInHierarchy(it) }
             || (type.superclass != null && hasAnnotationInHierarchy(type.superclass))
 }
+
+/**
+ * We can't use clazz.kotlin.getObjectInstance as that doesn't play nicely with nested private objects. Even
+ * setting the accessibility override (setAccessible) still causes an IllegalAccess Exception when attempting
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: IllegalAccessException (no space)

And quote (with square brackets?) the types / methods if you can.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt [hunk 2]
```diff
@@ -513,3 +513,34 @@ fun ClassWhitelist.hasAnnotationInHierarchy(type: Class<*>): Boolean {
...
+ * to retrieve the value of the INSTANCE field.
+ *
+ * Whichever reference to the class Kotlin reflection uses, override (set from setAccessible) on that field
+ * isn't set even when it was explicitly set as acceissible before calling into the kotlin reflection routines.
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: acceissible -> accessible.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt [hunk 3]
```diff
@@ -513,3 +513,34 @@ fun ClassWhitelist.hasAnnotationInHierarchy(type: Class<*>): Boolean {
...
+ *
+ * For example
+ *
+ * clazz.getDeclaredField("INSTANCE")?.apply {
+ *     isAccessible = true
+ *     kotlin.objectInstance // This throws as the INSTANCE field isn't accessible
+ * }
+ *
+ * Therefore default back to good old java reflection and simply look for the INSTANCE field as we are never going
+ * to serialize a companion object
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: full stop.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializerFactory.kt [hunk 4]
```diff
@@ -267,12 +267,16 @@ open class SerializerFactory(
...
                     if (clazz.componentType.isPrimitive) PrimArraySerializer.make(type, this)
                     else ArraySerializer.make(type, this)
-                } else if (clazz.kotlin.objectInstance != null) {
-                    whitelist.requireWhitelisted(clazz)
-                    SingletonSerializer(clazz, clazz.kotlin.objectInstance!!, this)
                 } else {
-                    whitelist.requireWhitelisted(type)
-                    ObjectSerializer(type, this)
+                   val singleton = clazz.objectInstance()
+
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializerFactory.kt]: remove blank line.

File: samples/simm-valuation-demo/src/main/kotlin/net/corda/vega/flows/SimmFlow.kt [hunk 5]
```diff
@@ -305,7 +305,8 @@ object SimmFlow {
         @Suspendable
         private fun agreePortfolio(portfolio: Portfolio): SignedTransaction {
             logger.info("Handshake finished, awaiting Simm offer")
-            require(offer.dealBeingOffered.portfolio == portfolio.refs)
+
+            require(offer.dealBeingOffered.portfolio.toSet().containsAll(portfolio.refs))
```
Comment (rick-r3) [inline samples/simm-valuation-demo/src/main/kotlin/net/corda/vega/flows/SimmFlow.kt]: Why is this change needed?  Went from same set to a super set.

Comment (fenryka) [reply samples/simm-valuation-demo/src/main/kotlin/net/corda/vega/flows/SimmFlow.kt]: List wasn't ordered, set ensures that doesn't matter

Comment (rick-r3) [reply samples/simm-valuation-demo/src/main/kotlin/net/corda/vega/flows/SimmFlow.kt]: Would be better to do `toSet()` on each and then use `==`.  Set equality ignores order, because it's a set, and doesn't really have order.  `containsAll` would allow a super set, which is different.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt (hunk 1 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: done

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt (hunk 2 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: done

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt (hunk 3 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: done

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializerFactory.kt (hunk 4 above)
Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializerFactory.kt]: I will... but  blank lines greatly help readability. 

In this case separating var declarations with actual code, it helps me find things legible but...

done

File: samples/simm-valuation-demo/src/main/kotlin/net/corda/vega/flows/SimmFlow.kt (hunk 5 above)
Comment (fenryka) [reply samples/simm-valuation-demo/src/main/kotlin/net/corda/vega/flows/SimmFlow.kt]: I was hoping for better set operations in Kotlin really, I'll change it

**Comment by fenryka** [comment]
@rick-r3 Changes made

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt [hunk 6]
```diff
@@ -513,3 +513,34 @@ fun ClassWhitelist.hasAnnotationInHierarchy(type: Class<*>): Boolean {
...
+ *     isAccessible = true
+ *     kotlin.objectInstance // This throws as the INSTANCE field isn't accessible
+ * }
+ *
+ * Therefore default back to good old java reflection and simply look for the INSTANCE field as we are never going
+ * to serialize a companion object.
+ */
+fun Class<*>.objectInstance() =
+    try {
+        this.getDeclaredField("INSTANCE")?.let { field ->
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: The `INSTANCE` field is static, right?  We should check that here.  Had to remind myself how to do it...

`field.modifiers` ANDed with `Modifiers.STATIC` != 0

Comment (rick-r3) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: I wonder what else we can do to reduce potential false positives?  Is the type of the field also the same as the `Class<*>` ?  That should be a check.  

Is there any other way to check if this is a kotlin `object` in the kotlin reflection API (without asking for the instance itself, which seems to be inaccessible)? I cannot see one.

Comment (rick-r3) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: The field is also `final`.

Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: every time you wander into Kotlin reflection it gets far more complicated than it should, so sticking with Java reflection for this wouldn't be a terrible thing, I may have gotten burnt trying to do this with Kotlin and wasting hours attempting it :)

Comment (rick-r3) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: And the class is also `final`.

Comment (rick-r3) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: And the class only has a `private` default (empty) constructor.

Comment (fenryka) [reply node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: done

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt [hunk 7]
```diff
@@ -515,9 +515,9 @@ fun ClassWhitelist.hasAnnotationInHierarchy(type: Class<*>): Boolean {
 }
 
 /**
- * We can't use clazz.kotlin.getObjectInstance as that doesn't play nicely with nested private objects. Even
- * setting the accessibility override (setAccessible) still causes an [IllegalAccessException] when attempting
- * to retrieve the value of the INSTANCE field.
+ * By default use Kotlin reflection and gran the objectInstance. However, that doesn't play nicely with nested
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/SerializationHelper.kt]: gran -> grab?
